### ISSUE TITLE
[REVIEW]: NGSTrefftz: Add-on to NGSolve for Trefftz methods

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#4135]
**Submitting author:** @PaulSt (<a href="http://orcid.org/0000-0001-5073-3366">Paul Stocker</a>)
**Repository:** https://github.com/PaulSt/NGSTrefftz
**Branch with paper.md** (empty if default branch): 
**Version:** v0.1.0
**Editor:** @Nikoleta-v3
**Reviewers:** @mscroggs, @thelfer 
**Archive:** 10.5281/zenodo.6394628

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/c2f4e85b118c22b81aa27d7799265409/status.svg)](https://joss.theoj.org/papers/c2f4e85b118c22b81aa27d7799265409)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/c2f4e85b118c22b81aa27d7799265409"><img src="https://joss.theoj.org/papers/c2f4e85b118c22b81aa27d7799265409/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/c2f4e85b118c22b81aa27d7799265409/status.svg)](https://joss.theoj.org/papers/c2f4e85b118c22b81aa27d7799265409)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/PaulSt/NGSTrefftz" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@mscroggs & @thelfer , please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @Nikoleta-v3 know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @mscroggs
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/PaulSt/NGSTrefftz">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/PaulSt/NGSTrefftz">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@PaulSt) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @thelfer 
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/PaulSt/NGSTrefftz">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/PaulSt/NGSTrefftz">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@PaulSt) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @mscroggs, @thelfer  it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 546

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.05 s (843.4 files/s, 177599.8 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
C++                             11            498            311           3260
C/C++ Header                     9            254             98           1158
Jupyter Notebook                11              0           1810            731
Python                           7            137            334            585
TeX                              1             14              0            119
YAML                             1              8              2            106
Markdown                         2             19              0             81
CMake                            1             10              5             54
Dockerfile                       2             15             15             40
JSON                             1              0              0             23
-------------------------------------------------------------------------------
SUM:                            46            955           2575           6157
-------------------------------------------------------------------------------


Statistical information for the repository '1d155acd4a973ae99ca7f710' was
gathered on 2022/02/07.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Christoph Lehrenfeld             1             2              2            0.00
Paul                           666         43191          36793           96.71
Paul Stocker                     5             9            442            0.55
PaulSt                          33          1392            871            2.74

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Christoph Lehrenfeld          2          100.0          0.0                0.00
Paul                       6633           15.4         15.1               10.58
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1016/j.camwa.2020.01.006 is OK
- 10.1007/s00211-017-0910-x is OK
- 10.1115/1.1995716 is OK
- 10.1016/0965-9978(95)00067-4 is OK
- 10.1007/978-3-319-41640-3_8 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.04135/joss.04135/10.21105.joss.04135.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.04135/joss.04135/10.21105.joss.04135.pdf) :page_facing_up: :point_left:

**Comment by Nikoleta-v3** [comment]
👋🏼  @PaulSt @mscroggs & @thelfer this is the review thread for the paper. All of our communications will happen here from now on.

Both reviewers have checklists at the top of this thread with the JOSS requirements 🔝 As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, the reviewers are encouraged to submit issues and pull requests on the software repository. When doing so, please mention https://github.com/openjournals/joss-reviews/issues/4135 so that a link is created to this thread (and I can keep an eye on what is happening). Please also feel free to comment and ask questions on this thread. In my experience, it is better to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

We aim for reviews to be completed within about 2-4 weeks. Please let me know if any of you require some more time. We can also use Whedon (our bot) to set automatic reminders if you know you'll be away for a known period of time.

Please feel free to ping me (@Nikoleta-v3 ) if you have any questions/concerns 👍🏻

**Comment by whedon** [comment]
:wave: @thelfer , please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @mscroggs, please update us on how your review is going (this is an automated reminder).

**Comment by thelfer** [comment]
I currently have a working version of `NGSTrefftz`, after resolution of various issues. Review starts now !

**Comment by mscroggs** [comment]
I've been busier than I expect the last two weeks. My review will be done next week

**Comment by thelfer** [comment]
@Nikoleta-v3 Would you remind me what I need to do to be able to check my checklist ? Sorry if I missed something ?

**Comment by thelfer** [comment]
@editorialbot generate my checklist

**Comment by thelfer** [comment]
@Nikoleta-v3. All my questions and comments have been answered by @PaulSt (See Issues 2 to 9). I consider my review as over and would recommend the publication of the paper.

I think that we can congratulate @PaulSt for his patience.

**Comment by PaulSt** [comment]
Thank you again @thelfer for the detailed review and very helpful feedback!

**Comment by mscroggs** [comment]
@PaulSt Nice software and nice paper!

I've opened a PR (https://github.com/PaulSt/NGSTrefftz/pull/10) to remove some brackets around references and add a reference to the paper that the FreeHyTe people wrote. Once that PR is resolve, I think this is ready to be accepted

**Comment by Nikoleta-v3** [comment]
Thank you @thelfer and @mscroggs for your time and your reviews! @PaulSt once you have addressed the reviewers comments I am also going to have a final look 👍🏻

**Comment by mscroggs** [comment]
@Nikoleta-v3 My comments are all now dealt with and I've ticked all the boxes for my review

**Comment by PaulSt** [comment]
Thank you for your PR and the review @mscroggs!

**Comment by Nikoleta-v3** [comment]
Again a huge thank you to the reviewers!

**Comment by thelfer** [comment]
> Again a huge thank you to the reviewers!

My pleasure to contribute to JOSS.

**Comment by Nikoleta-v3** [comment]
@editorialbot generate pdf

**Comment by Nikoleta-v3** [comment]
@editorialbot generate pdf

**Comment by Nikoleta-v3** [comment]
@whedon generate pdf

**Comment by Nikoleta-v3** [comment]
@editorialbot generate pdf

**Comment by Nikoleta-v3** [comment]
Hey @PaulSt 👋🏻 Could you please have a look at the following (minor) issues?

### Documentation

On the NGSTrefftz [website](https://paulst.github.io/NGSTrefftz/):

1. The hyperlinks to NGSolve do not work
2. The emojis `:tada::+1:` are not rendered 

### Readme

At the installation section you are listing the dependencies and their versions :+1: Could you include hyperlinks to the software’s website, or their installation pages?

This is more of a suggestion: why don’t you make `Papers using the code section` a section instead of a subsection?

### Testing

I can see you have some tests (folder `test`) but I couldn't find any information on how to run the tests. Can you include this information in the documentation? And please point to it in, or just include it under `Contributing to NGSTrefftz`. 

I can also see that you are running GitHub actions. Could you also run the test suite as part of the workflow?

**Comment by Nikoleta-v3** [comment]
Thank you @PaulSt. Note that you left some commented out [lines](https://github.com/PaulSt/NGSTrefftz/blob/ffe8537814137f59bb0f4d0a31d2527f9b78042e/.github/workflows/build.yml#L75-L81) in the `build.yml` file.

At this point could you:

- [ ] Let me know which tagged release of your software we should use (I am assuming [v0.0.7](https://github.com/PaulSt/NGSTrefftz/releases/tag/v0.0.7)).
- [ ] Archive the reviewed software in Zenodo or a similar service (e.g., figshare, an institutional repository)
- [ ] Check the archival deposit (e.g., in Zenodo) has the correct metadata. This includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it). You may also add the authors' ORCID.
- [ ] Please list the DOI of the archived version here.

I can then move forward with accepting the submission. Let me know if you need any help with any of the above 😄

**Comment by PaulSt** [comment]
Hi @Nikoleta-v3, 
and sorry for the delay! 
- [X] I published a release which I would like to use for the software, https://github.com/PaulSt/NGSTrefftz/releases/tag/v0.1.0
- [X] The software is published in Zenodo at [10.5281/zenodo.6394628](https://zenodo.org/record/6394628)

**Comment by Nikoleta-v3** [comment]
@editorialbot set [10.5281/zenodo.6394628](https://zenodo.org/record/6394628) as archive

**Comment by Nikoleta-v3** [comment]
@editorialbot set v0.1.0 as version

**Comment by Nikoleta-v3** [comment]
@editorialbot recommend-accept

**Comment by PaulSt** [comment]
Hi @Nikoleta-v3,
I was just checking if the error was an issue in the paper but locally it compiled. Is it possible that 
> @editorialbot set [10.5281/zenodo.6394628](https://zenodo.org/record/6394628) as archive

loops back onto itself? So one would need to call it without the link? :thinking:

**Comment by Nikoleta-v3** [comment]
Thank you for checking the error! Let me try that 😄

**Comment by Nikoleta-v3** [comment]
@editorialbot set 10.5281/zenodo.6394628 as archive

**Comment by Nikoleta-v3** [comment]
@editorialbot recommend-accept

**Comment by Nikoleta-v3** [comment]
Thank you @PaulSt for helping with the compiling issue 😄

**Comment by PaulSt** [comment]
> 
> Check final proof point_right [openjournals/joss-papers#3099](https://github.com/openjournals/joss-papers/pull/3099)
> 
@Nikoleta-v3 is missing as the editor in the proof, everything else looks good to me!

**Comment by kyleniemeyer** [comment]
@openjournals/dev I think the editor should be showing up properly in the proof, any idea what's going on there?

**Comment by kyleniemeyer** [comment]
The proof looks good otherwise, but I'll wait for that issue to be fixed to publish.

**Comment by xuanxu** [comment]
> @openjournals/dev I think the editor should be showing up properly in the proof, any idea what's going on there?

Probably this can be fixed by @Nikoleta-v3 filling in *first_name/last_name* fields in her [editor profile page](https://joss.theoj.org/editor_profile).

**Comment by Nikoleta-v3** [comment]
Yeah they are already filled. I've edited other submissions in the past and my name appeared without a problem 😄

**Comment by xuanxu** [comment]
Ouch, then sounds like there's a bug somewhere in the code preparing the paper. I'm going to try to locate it.

**Comment by xuanxu** [comment]
Bug found, it was caused by incorrectly managing usernames with upcased letters. 
Fix applied, regenerating the paper should get the right editor info now.

**Comment by Nikoleta-v3** [comment]
Thank you very much @xuanxu!

Should I run the `generate pdf command`? will this update the final proofs or should I re-run `@editorialbot recommend-accept`?

**Comment by xuanxu** [comment]
@Nikoleta-v3 yeah, let's try the `recommend-accept` again🤞

**Comment by Nikoleta-v3** [comment]
@editorialbot recommend-accept

**Comment by xuanxu** [comment]
Yay! 🎉 
![Captura de pantalla 2022-03-31 a las 13 21 20](https://user-images.githubusercontent.com/6528/161043946-2fb657cc-cc05-474c-a06d-9f6ffe354b3b.png)

**Comment by Nikoleta-v3** [comment]
I am not invisible anymore 👻 🎉

**Comment by kyleniemeyer** [comment]
@editorialbot accept

**Comment by kyleniemeyer** [comment]
Congratulations @PaulSt on your article's publication in JOSS!

Many thanks to @mscroggs and @thelfer for reviewing this, and @Nikoleta-v3 for editing.

**Comment by PaulSt** [comment]
Hi @kyleniemeyer,

thank you very much and many thanks to @Nikoleta-v3 for editing! 

I was looking at the publication and was surprised to see the quotation mark ` used to style the title in markdown made it into the title on the joss homepage. I don't know how that happened, and my apologies if this was an error on my side. If still possible, it might be nicer to remove it, to make it more uniform with the other papers.

**Comment by kyleniemeyer** [comment]
@PaulSt ah, good catch. that is probably not supposed to happen.

@openjournals/dev is this something that can be fixed?

<img width="1074" alt="Screen Shot 2022-03-31 at 6 12 10 PM" src="https://user-images.githubusercontent.com/190432/161175665-d60ec6cb-12e7-42c8-9b94-ab9667e4ed93.png">

**Comment by PaulSt** [comment]
Hi :wave: @kyleniemeyer any update on this? :thinking: 
> 
> @openjournals/dev is this something that can be fixed?
>
